File: python/medium/Pow(x, n).py
class Solution:
    def Pow(self, x: float, n: int) -> float:
        if n == 1:
            return x
        elif n % 2 == 0:
            return self.Pow(x*x, abs(n)/2)
        else:
            return x*self.Pow(x*x, (abs(n)-1)/2)
    
    def myPow(self, x: float, n: int) -> float:
        # return x**n
        if n == 0:
            return 1
        answer = self.Pow(x, abs(n))
        if n > 0:
            return answer
        else:
            return 1/answer

File: cpp/medium/Pow(x, n).cpp
class Solution {
public:
    double myPow(double x, int n) {
        if (n < 0) {
            n = abs(n);
            x = 1 / x;
        }
        double answer = 1;
        while (n > 0) {
            if (n % 2 != 0) {
                answer*=x;
            }
            x*=x;
            // bit位移會快很多
            // n/=2;
            n >>= 1;
        }
        return answer;
    }
};
